test/extensions/filters/listener/common/fuzz/listener_filter_fuzzer.cc

```diff
@@ -10,21 +8,86 @@ void ListenerFilterFuzzer::fuzz(
     Network::ListenerFilter& filter,
     const test::extensions::filters::listener::FilterFuzzTestCase& input) {
   try {
-    fuzzerSetup(input);
+    socket_.setLocalAddress(Network::Utility::resolveUrl(input.sock().local_address()));
   } catch (const EnvoyException& e) {
-    ENVOY_LOG_MISC(debug, "EnvoyException: {}", e.what());
-    return;
+    // Socket's local address will be nullptr by default if fuzzed local address is malformed
+    // or missing - local address field in proto is optional
+  }
+  try {
+    socket_.setRemoteAddress(Network::Utility::resolveUrl(input.sock().remote_address()));
+  } catch (const EnvoyException& e) {
+    // Socket's remote address will be nullptr by default if fuzzed remote address is malformed
+    // or missing - remote address field in proto is optional
+  }
+
+  FuzzedHeader header(input);
+
+  if (!header.empty()) {
+    ON_CALL(os_sys_calls_, recv(kFakeSocketFd, _, _, MSG_PEEK))
+        .WillByDefault(testing::Return(Api::SysCallSizeResult{static_cast<ssize_t>(0), 0}));
+
+    ON_CALL(dispatcher_,
+            createFileEvent_(_, _, Event::FileTriggerType::Edge,
+                             Event::FileReadyType::Read | Event::FileReadyType::Closed))
+        .WillByDefault(testing::DoAll(testing::SaveArg<1>(&file_event_callback_),
+                                      testing::ReturnNew<NiceMock<Event::MockFileEvent>>()));
   }
 
   filter.onAccept(cb_);
+
+  if (file_event_callback_ == nullptr) {
+    // If filter does not call createFileEvent (i.e. original_dst and original_src)
+    return;
+  }
+
+  if (!header.empty()) {
+    {
+      testing::InSequence s;
+
+      EXPECT_CALL(os_sys_calls_, recv(kFakeSocketFd, _, _, MSG_PEEK))
+          .Times(testing::AnyNumber())
+          .WillRepeatedly(Invoke([&header](os_fd_t, void* buffer, size_t length, int)
+                                 -> Api::SysCallSizeResult {
+            return header.next(buffer, length);
+          }));
+    }
+
+    bool got_continue = false;
+
+    ON_CALL(cb_, continueFilterChain(true))
+        .WillByDefault(testing::InvokeWithoutArgs([&got_continue]() { got_continue = true; }));
+
+    while (!got_continue) {
+      if (header.done()) { // End of stream reached but not done
+        file_event_callback_(Event::FileReadyType::Closed);
+      } else {
+        file_event_callback_(Event::FileReadyType::Read);
+      }
+    }
+  }
 }
 
-void ListenerFilterFuzzer::socketSetup(
-    const test::extensions::filters::listener::FilterFuzzTestCase& input) {
-  socket_.setLocalAddress(Network::Utility::resolveUrl(input.sock().local_address()));
-  socket_.setRemoteAddress(Network::Utility::resolveUrl(input.sock().remote_address()));
+FuzzedHeader::FuzzedHeader(const test::extensions::filters::listener::FilterFuzzTestCase& input)
+    : nreads(input.data_size()), nread(0), header("") {
+  for (int i = 0; i < nreads; i++) {
+    header += input.data(i);
```

Comment:
This loop takes O(n^2) time since the memory for the string might need to be reallocated in every iteration. Consider either
1) reserving capacity using `std::string::reserve()` to allocate the full length up front, or
2) saving each part of the string individually, and copying up to `nread` of them in `next()`

Reply:
Chose to implement (1) with an additional for loop to calculate the full length of the header,